### PR TITLE
WI #1641 Fix generated code when a procedure uses type defined in dep…

### LINKED ISSUE #1641: Wrong generated code with multiple COPY instructions
**Describe the bug**
In some cases Codegen will output multiple COPY instructions instead of one. The generated code won't compile because of multiple equivalent variable declarations.

**To Reproduce**
Consider a copybook `TESTCPY.cpy`, a dependency `Library.tcbl` and a main file `Client.tcbl`.

`TESTCPY.cpy`
```cobol
             10 varA-1 pic X.
             10 varA-2.
                15 varB-1 pic X.
             10 varA-3 pic X.
```

`Library.tcbl`
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. Library.
       DATA DIVISION.
       WORKING-STORAGE SECTION.                                               
       01 Type1 typedef strict public.
          05  Body. COPY testcpy.
       END PROGRAM Library.

```

`Client.tcbl`
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. Client.
      
       PROCEDURE DIVISION.
      
       declare procedure MyProc
       input myType1 type Library::Type1.
       PROCEDURE DIVISION.
           DISPLAY myType1
           .
       end-declare.
      
       END PROGRAM Client.
```

The generated code will then contain in linkage section :
```cobol
       01 myType1.
           02 Body.
          COPY testcpy.
          COPY testcpy.
```

**Expected behavior**
Same generation but with only one COPY instruction.

Expected (Type)Cobol code : 
```cobol
       01 myType1.
           02 Body.
          COPY testcpy.
```

**Note** : the problem seems to be located in `TypedDataNode.InsertChildren` method. The `usedCopies` collection is cleared during recursive call and the COPY instruction is written multiple times.

### PR DESCRIPTION
**Pull request by LANOYDO**
…endency and copies

Fixes #1641.

### CODE REVIEW

File: CLI/test/ressources/CopyInsideTypeDefInsideDependency/input/Copies/YCONVEA.CPY [hunk 1]
```diff
@@ -1,6 +1,7 @@
          05  CONVEA-VEA-VEA-OUT-VEA pic X.
          05  CONVEA-VEA-VEA-OUT-VEA-2 pic X.
-         05  CONVEA pic X.
+         05  CONVEA.
+                10 CONVEA-SUB pic X.
```
Comment (fm-117) [inline CLI/test/ressources/CopyInsideTypeDefInsideDependency/input/Copies/YCONVEA.CPY]: Could you add another sublevel either in this copy or in the other one ?

File: CLI/test/ressources/CopyInsideTypeDefInsideDependency_EI/input/Copies/YCONVEA.CPY [hunk 2]
```diff
@@ -1,6 +1,7 @@
          05  CONVEA-VEA-VEA-OUT-VEA pic X.
          05  CONVEA-VEA-VEA-OUT-VEA-2 pic X.
-         05  CONVEA pic X.
+         05  CONVEA.
```
Comment (fm-117) [inline CLI/test/ressources/CopyInsideTypeDefInsideDependency_EI/input/Copies/YCONVEA.CPY]: Same remark as in the non-EI test.

File: Codegen/src/Nodes/TypedDataNode.cs [hunk 3]
```diff
@@ -972,11 +972,15 @@ private static System.Text.StringBuilder GetIndent(int level, int indent, bool i
 
         };
 
-        public static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string> rootProcedures, List< Tuple<string,string> > rootVariableName, DataDefinition ownerDefinition, DataDefinition type, int level, int indent)
+        public static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string> rootProcedures, List<Tuple<string, string>> rootVariableName, DataDefinition ownerDefinition, DataDefinition type, int level, int indent)
```
Comment (fm-117) [inline Codegen/src/Nodes/TypedDataNode.cs]: Considering all the usages of the public method, you can merge `DataDefinition ownerDefinition` and `DataDefinition type` as a single parameter and also remove `int indent` parameter.

File: Codegen/src/Nodes/TypedDataNode.cs [hunk 4]
```diff
@@ -1002,10 +1006,10 @@ public static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string>
                         CopyDirective copy = child.CodeElement.FirstCopyDirective;
                         //The first data coming from a copy is used to recover the Clause COPY, the other would be only a repetition of this one, so we skip them.
                         //Even with the same name, two different Clause COPY are differentiated by their token lines
-                        if (usedCopies.Contains(copy) == false)
+                        if (usedCopy == null || usedCopy.Equals(copy) == false)
```
Comment (fm-117) [inline Codegen/src/Nodes/TypedDataNode.cs]: Why use `Equals` here ? `CopyDirective` does not define custom equality comparison mechanism so reference comparison is fine.

File: Codegen/src/Nodes/TypedDataNode.cs [hunk 5]
```diff
@@ -972,11 +972,15 @@ private static System.Text.StringBuilder GetIndent(int level, int indent, bool i
 
         };
 
-        public static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string> rootProcedures, List< Tuple<string,string> > rootVariableName, DataDefinition ownerDefinition, DataDefinition type, int level, int indent)
+        public static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string> rootProcedures, List<Tuple<string, string>> rootVariableName, DataDefinition ownerDefinition, DataDefinition type, int level, int indent)
+        {
+            return InsertChildren(layout, rootProcedures, rootVariableName, ownerDefinition, type, level, indent, null);
+        }
+
+        private static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string> rootProcedures, List< Tuple<string,string> > rootVariableName, DataDefinition ownerDefinition, DataDefinition type, int level, int indent, CopyDirective usedCopy)
```
Comment (fm-117) [inline Codegen/src/Nodes/TypedDataNode.cs]: So here you add a new `CopyDirective usedCopy` parameter to track the current COPY instruction and to know whether it should be written or not. But @smedilol confirmed that COPY instructions may be found in copybook content so could you please add a test having a COPY within a COPY ? I think we may need a `Stack<CopyDirective>` here.

File: CLI/test/ressources/CopyInsideTypeDefInsideDependency/input/Copies/YCONVEA.CPY (hunk 1 above)
Comment (LANOYDO) [reply CLI/test/ressources/CopyInsideTypeDefInsideDependency/input/Copies/YCONVEA.CPY]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1646/commits/5ec741f218377464909aae42e8454ea4de4d1d2e

File: CLI/test/ressources/CopyInsideTypeDefInsideDependency_EI/input/Copies/YCONVEA.CPY (hunk 2 above)
Comment (LANOYDO) [reply CLI/test/ressources/CopyInsideTypeDefInsideDependency_EI/input/Copies/YCONVEA.CPY]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1646/commits/5ec741f218377464909aae42e8454ea4de4d1d2e

File: Codegen/src/Nodes/TypedDataNode.cs (hunk 3 above)
Comment (LANOYDO) [reply Codegen/src/Nodes/TypedDataNode.cs]: It's right; done.
https://github.com/TypeCobolTeam/TypeCobol/pull/1646/commits/5ec741f218377464909aae42e8454ea4de4d1d2e

File: Codegen/src/Nodes/TypedDataNode.cs (hunk 4 above)
Comment (LANOYDO) [reply Codegen/src/Nodes/TypedDataNode.cs]: using ==; done.
https://github.com/TypeCobolTeam/TypeCobol/pull/1646/commits/5ec741f218377464909aae42e8454ea4de4d1d2e

File: Codegen/src/Nodes/TypedDataNode.cs (hunk 5 above)
Comment (LANOYDO) [reply Codegen/src/Nodes/TypedDataNode.cs]: - Test completed with adding use of copy of copy (CopyInsideTypeDefInsideDependency).
- In method InsertChildren, when a copy is found, the parsing of nodes is interrompted by instruction "continue" (line 1013); the copy below the copy is not ultimately considered.
https://github.com/TypeCobolTeam/TypeCobol/pull/1646/commits/5ec741f218377464909aae42e8454ea4de4d1d2e

File: Codegen/src/Nodes/TypedDataNode.cs [hunk 6]
```diff
@@ -972,11 +972,15 @@ private static System.Text.StringBuilder GetIndent(int level, int indent, bool i
 
         };
 
-        public static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string> rootProcedures, List< Tuple<string,string> > rootVariableName, DataDefinition ownerDefinition, DataDefinition type, int level, int indent)
+        public static List<ITextLine> InsertChildren(ColumnsLayout? layout, List<string> rootProcedures, List<Tuple<string, string>> rootVariableName, DataDefinition type, int level)
```
Comment (fm-117) [inline Codegen/src/Nodes/TypedDataNode.cs]: The `type` parameter can be more strictly typed as `TypeDefinition` here.

**Pull request merged**